### DISCUSSION TITLE
Does RESTEasy support EJB 2.0 beans?

### DISCUSSION THREAD

**Discussion by tituslup** [resteasy/resteasy#4150] [Q&A]
Hi,

Does RESTEasy support EJB 2.0 beans?

RESTEasy 6.2.5 documentation points to using annotations for interfaces and beans, but in my application this is handled using ejb-jar.xml descriptor files, for example:
```xml
<session>
	<display-name>FolderService</display-name>
	<ejb-name>FolderService</ejb-name>
	<home>com.example.folder.ejb.session.FolderServiceHome</home>
	<remote>com.example.folder.FolderService</remote>
	<ejb-class>com.example.folder.ejb.session.FolderServiceBean</ejb-class>
	<session-type>Stateless</session-type>
	<transaction-type>Container</transaction-type>
</session>
```

Thanks,
-Titus

**Comment by jamezp** [comment]
RESTEasy 6.2.5 will not support EJB 2.0 as it targets Jakarta EE 10. It will support EJB's however https://jakarta.ee/specifications/restful-ws/3.1/jakarta-restful-ws-spec-3.1#async_ejbs.

**Reply by tituslup** [reply]
Thanks for answer.

So my options are either use EJB resource classes (POJO like) or migrate my EJB2 classes to fit the pattern described here [RESTEeasy EJB Integration](https://docs.jboss.org/resteasy/docs/6.2.5.Final/userguide/html_single/index.html#RESTEasy_Enterprise_Bean_Integration)?
Is there any difference with respect to that manual JNDI resource mapping in web.xml? Is that something that is not required for EJB resource classes?

**Reply by jamezp** [reply]
I do understand what  mean by EJB 2 now. You just mean defining the types in a `ejb.xml`. According to the spec you can use an XML descriptor https://jakarta.ee/specifications/enterprise-beans/4.0/jakarta-enterprise-beans-spec-core-4.0#a5910.

It may work in WildFly, but I've never tried it. I'd have to try to create an example project to try it. I'm not sure when I'll have time to do that unfortunately.

**Reply by tituslup** [reply]
Its not only the the structural information that is mentioned in the ejb-jar.xml, but the classes involved are tied to the EJB objects, something that has changed and simplified in EJB3 specification.

I have tried to create an example based on my application code, an ear file with two ejb jar files, one with the old EJB 2 approach and one with EJB 3:
* well EJB2 approach was problematic and detailed below;
* EJB3 approach bean inheriting the interface, resource detection worked and endpoint was directly exposed and working;
* both situations wildfly detected the beans and management portal shows both types of beans.

The problem I'm facing is there is no resource detection for EJB2 approach and manually adding the resources in web.xml just leads to all kind of RESTEASY errors, which I'm unsure how to handle. Adding the interface as a resource in web.xml leads to **RESTEASY003190: Could not find constructor for class**. Adding the bean as a resource leads to  **RESTEASY003130: Class is not a root resource** as only the interface is annotated with Path, if I add annotation to the bean that obviously fails because **@Path annotations can only be placed on classes or interfaces that represent a local, remote or no-interface view** and so on.
Seems like resource detection would be the way to go.

Interface
```java
package com.example.oldejb;

import jakarta.ejb.EJBObject;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.Path;

@Path("/ejb2")
public interface TestEJB2Interface extends EJBObject {
    @GET
    public String sayHelloFromServiceBean();
}
```

Bean
```java
package com.example.oldejb;
import jakarta.ejb.SessionBean;
import jakarta.ejb.EJBException;
import jakarta.ejb.CreateException;
import jakarta.ejb.SessionContext;
import java.rmi.RemoteException;
import javax.security.auth.callback.CallbackHandler;


public class TestEJB2Bean implements SessionBean
{
    public String sayHelloFromServiceBean() {
        System.out.println("Hello From Service Bean EJB 2!");
        return "Hi EJB 2";
    }

    public void setSessionContext(SessionContext ctx) {
    }

    public void ejbRemove() throws EJBException, RemoteException {
    }

    public void ejbActivate() throws EJBException, RemoteException {
    }

    public void ejbPassivate() throws EJBException, RemoteException {
    }

    public void ejbCreate() throws CreateException {
    }
}
```

Home
```java
package com.example.oldejb;
import com.example.oldejb.TestEJB2Interface;
import jakarta.ejb.CreateException;
import jakarta.ejb.EJBHome;

public interface TestEJB2Home extends EJBHome {
    TestEJB2Interface create() throws CreateException;
```

ejb-jar.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE ejb-jar PUBLIC "-//Sun Microsystems, Inc.//DTD Enterprise JavaBeans 2.0//EN" "http://java.sun.com/dtd/ejb-jar_2_0.dtd">
<ejb-jar>
    <enterprise-beans>
        <session>
            <display-name>TestEJB2Interface</display-name>
            <ejb-name>TestEJB2Interface</ejb-name>
            <home>com.example.oldejb.TestEJB2Home</home>
            <remote>com.example.oldejb.TestEJB2Interface</remote>
            <ejb-class>com.example.oldejb.TestEJB2Bean</ejb-class>
            <session-type>Stateless</session-type>
            <transaction-type>Container</transaction-type>
        </session>
    </enterprise-beans>
</ejb-jar>
```

web.xml
```xml
<web-app version="3.0" xmlns="http://java.sun.com/xml/ns/javaee"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd">
    <display-name>restapi</display-name>
    <description>REST interface</description>
    <context-param>
        <param-name>resteasy.server.tracing.type</param-name>
        <param-value>ALL</param-value>
    </context-param>
    <context-param>
        <param-name>resteasy.server.tracing.threshold</param-name>
        <param-value>VERBOSE</param-value>
    </context-param>
    <context-param>
        <param-name>resteasy.jndi.resources</param-name>
        <param-value>java:global/restapi/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Interface</param-value>
    </context-param>
</web-app>
```
This JDNI resource mapping I'm not so sure what should be the value. Pretty much tried it with every mapping reported by wildfly for the application.
```
java:global/restapi/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Home
java:app/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Home
java:module/TestEJB2Interface!com.example.oldejb.TestEJB2Home
java:global/restapi/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Interface
java:app/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Interface
java:module/TestEJB2Interface!com.example.oldejb.TestEJB2Interface
java:jboss/exported/restapi/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Interface
ejb:restapi/ejb2/TestEJB2Interface!com.example.oldejb.TestEJB2Interface
```

application.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<application xmlns="http://java.sun.com/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
             version="6"
             xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/application_6.xsd">
  <application-name>restapi</application-name>
  <module>
    <web>
      <web-uri>restapi.war</web-uri>
      <context-root>restapi</context-root>
    </web>
  </module>
  <module>
    <ejb>ejb.jar</ejb>
  </module>
  <module>
    <ejb>ejb2.jar</ejb>
  </module>
</application>
```

deployment structure
```
<jboss-deployment-structure>
    <deployment name="restapi.war">
        <dependencies>
            <module name="jakarta.ws.rs.api" services="import"/>
            <module name="jakarta.ejb.api"/>
            <module name="org.jboss.ejb-client"/>
        </dependencies>
        <resources>
            <resource-root path="ejb.jar" />
            <resource-root path="ejb2.jar" />
        </resources>
    </deployment>
</jboss-deployment-structure>
```

**Reply by jamezp** [reply]
Just to let you know, I'm not ignoring this. I'm on PTO the Wednesday and Thursday and just haven't had a chance to look at this. I apologize, but I've not forgotten :)

**Reply by tituslup** [reply]
I've been playing around some more with this, trying to change the various versions at the top of the xmls, from 2 to 3.1, to exclude EJB2 from the equation, for example:
```xml
<ejb-jar xmlns="http://java.sun.com/xml/ns/javaee"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://java.sun.com/xml/ns/javaee
          http://java.sun.com/xml/ns/javaee/ejb-jar_3_1.xsd"
         version="3.1">
...
```

```xml
<application xmlns="http://xmlns.jcp.org/xml/ns/javaee"
             xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
             xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee
        http://xmlns.jcp.org/xml/ns/javaee/application_7.xsd"
             version="7">
...
```
No luck. Wildfly loads it, EJB JNDI mappings are displayed in the deployment logs and EJB is listed in wildfly management interface, but JAX-RS does not show it as a REST Resource, end point does not work.

Seems like the more appropriate question is if RESTEasy supports EJB2 style beans, where all information is captured in an XML descriptor, which is still a valid use case in EJB4 specifications, if I understand this correctly.

If you have any luck with your example, please let me know.
Or if you can suggest what do try next, I can give it a shot.

**Reply by jamezp** [reply]
Sorry, I keep getting sucked into random things and haven't had a chance to look at this. RESTEasy itself doesn't do much with EJB's. It relies on the CDI container really. That said in the RESTEasy CDI Extension, there are some specific checks for stateless and singleton beans. All that does though is stop the bean from getting registered. However, it explicitly looks for the annotation.

There is a chance today I can attempt to create a reproducer and see what happens.

**Reply by jamezp** [reply]
I wasn't able to create a reproducer as it's quite involved, for me at least as I'm not an EJB expert :), but I did see this in the [spec](https://jakarta.ee/specifications/restful-ws/3.1/jakarta-restful-ws-spec-3.1#ejbs):
> Resource class annotations (like `@Path`) MUST be applied to an EJB’s class directly following the annotation inheritance rules defined in [Annotation Inheritance](https://jakarta.ee/specifications/restful-ws/3.1/jakarta-restful-ws-spec-3.1#annotationinheritance).

**Reply by jamezp** [reply]
I used slightly different names, but if I annotate `TestEJB2Bean` with the Jakarta REST annotations, all seems to work. From the spec link above, I think that is required.

**Reply by tituslup** [reply]
Just to clarify, the spec link points to inheriting directly from the interface and using the EJB annotations (the new EJB3 approach) and no descriptor file. Are you saying in your example you are using this approach, that it worked this way and this is the only way it can work?

1. If I add @Remote to my interface (the one inheriting from EJBObject) and @Stateless to the bean class (the one inheriting from SessionBean), deployment works, but nothing happens with respect to exposing the endpoint.
2. If I add @Path as well to the bean, deployment fails with:
```
Jakarta RESTful Web Services resource [class com.example.oldejb.TestEJB2Bean] does not correspond to a view on the Jakarta Enterprise Beans TestEJB2Interface. @Path annotations can only be placed on classes or interfaces that represent a local, remote or no-interface view of an Jakarta Enterprise Beans.
```
3. If I drop the descriptor entries, EJBObject and SessionBean (in addition to 1 and 2), essentially using the EJB3 approach, then in works.

**Reply by jamezp** [reply]
All I did was add the `@Path` and `@GET` annotations to the `TestEJB2Bean`. I did not add EJB annotations. I've attached my example project.

[ejb-in-jar.zip](https://github.com/resteasy/resteasy/files/15310052/ejb-in-jar.zip)

**Reply by tituslup** [reply]
Thank you for providing the example.
I've ran 'mvn clean install' to generate the packages and copied ejb-war.war to my deployments folder to deploy it.
Deployment worked, management console JAX-RS shows the REST resource and end point indeed works.
```
19:03:37,099 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-6) WFLYSRV0027: Starting deployment of "ejb-war.war" (runtime-name: "ejb-war.war")
19:03:37,301 INFO  [org.jboss.weld.deployer] (MSC service thread 1-5) WFLYWELD0003: Processing weld deployment ejb-war.war
19:03:37,587 INFO  [org.jboss.weld.Version] (MSC service thread 1-5) WELD-000900: 5.1.2 (Final)
19:03:38,562 INFO  [org.jboss.resteasy.resteasy_jaxrs.i18n] (ServerService Thread Pool -- 5) RESTEASY002225: Deploying jakarta.ws.rs.core.Application: class dev.resteasy.examples.web.rest.RestActivator$Proxy$_$$_WeldClientProxy
19:03:38,572 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 5) WFLYUT0021: Registered web context: '/ejb-war' for server 'default-server'
19:03:38,637 INFO  [org.jboss.as.server] (DeploymentScanner-threads - 1) WFLYSRV0010: Deployed "ejb-war.war" (runtime-name : "ejb-war.war")
```


The only problems I see:
* During deployment time there were no JNDI binding logs for this bean
* Management console does not list GreetingsService* under runtime->EJB section.

I wonder what this means ... if this EJB is valid or not.
  
  

Just to see what happens, I've copied ejb-jar.jar under the deployments folder and attempted to deploy that, JNDI bindings were logged this time, but deployment failed with the previously mentioned error:
```
2024-05-14 18:06:47,764 INFO Thread-17 [org.jboss.as.server.deployment.AbstractDeploymentUnitService] [start] WFLYSRV0027: Starting deployment of "ejb-jar.jar" (runtime-name: "ejb-jar.jar")
2024-05-14 18:06:47,789 INFO Thread-23 [org.jboss.as.ejb3.deployment.processors.EjbJndiBindingsDeploymentUnitProcessor] [setupJNDIBindings] WFLYEJB0473: JNDI bindings for session bean named 'GreetingService' in deployment unit 'deployment "ejb-jar.jar"' are as follows:

        java:global/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingServiceHome
        java:app/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingServiceHome
        java:module/GreetingService!dev.resteasy.examples.ejb.service.GreetingServiceHome
        java:jboss/exported/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingServiceHome
        java:global/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingService
        java:app/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingService
        java:module/GreetingService!dev.resteasy.examples.ejb.service.GreetingService
        java:jboss/exported/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingService
        ejb:/ejb-jar/GreetingService!dev.resteasy.examples.ejb.service.GreetingService

2024-05-14 18:06:47,790 ERROR Thread-23 [org.jboss.msc.service.ServiceControllerImpl] [startFailed] MSC000001: Failed to start service jboss.deployment.unit."ejb-jar.jar".POST_MODULE
org.jboss.msc.service.StartException in service jboss.deployment.unit."ejb-jar.jar".POST_MODULE: WFLYSRV0153: Failed to process phase POST_MODULE of deployment "ejb-jar.jar"
        at org.jboss.as.server@22.0.1.Final//org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:178)
        at org.jboss.msc@1.5.2.Final//org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1617)
        at org.jboss.msc@1.5.2.Final//org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1580)
        at org.jboss.msc@1.5.2.Final//org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1438)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1990)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
        at org.jboss.threads@2.4.0.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.jboss.as.server.deployment.DeploymentUnitProcessingException: WFLYRS0010: Jakarta RESTful Web Services resource [class dev.resteasy.examples.ejb.service.GreetServiceSessionBean] does not correspond to a view on the Jakarta Enterprise Beans GreetingService. @Path annotations can only be placed on classes or interfaces that represent a local, remote or no-interface view of an Jakarta Enterprise Beans.

```

**Reply by jamezp** [reply]
Ah, good catch! It looks like my deployment is wrong. Let me see if I can fix that.

**Comment by tituslup** [comment]
Just to close the loop on this thread, I wasn't able to find a working solution using the xml descriptor that ties all the interfaces/beans/home classes. Instead I've focused on migrating to the EJB 3.X structure, with annotations and with no mapping inside the descriptor xml. More code changes, but definitely works.
@jamezp thanks for the inputs.

**Reply by jamezp** [reply]
Sorry about that. I got distracted on other tasks and didn't find time to pick this back up.
